server/src/main/java/com/ncc/neon/services/ClusterService.java

```diff
@@ -23,16 +26,190 @@
     /**
      * Clusters the results from the query.
      *
-     * TODO: Currently returns the same results i.e. method is a passthrough right now
-     *
      * @param tabularQueryResult the result of the unclustered query
      * @return the new clustered results
      */
     public TabularQueryResult cluster(TabularQueryResult tabularQueryResult) {
+        if (this.clusterClause.getFieldType().equals("number")) {
+            return this.aggregateNumber(tabularQueryResult);
+        }
+
         this.clusterClause = null; // reset cluster clause
         return tabularQueryResult;
     }
 
+    /**
+     * Clusters results by number aggregation.
+     *
+     * @param tabularQueryResult the original results
+     * @return the tabularqueryresult aggregated by numbers with the specifications given by the cluster clause
+     */
+    private TabularQueryResult aggregateNumber(TabularQueryResult tabularQueryResult) {
+        List<List<Object>> clusters = this.clusterClause.getClusters();
+
+        // establish keys
+        String fieldNameKey = this.clusterClause.getFieldNames().get(0);
+        String aggregationNameKey = this.clusterClause.getAggregationName();
+
+        List<Map<String, Object>> data = tabularQueryResult.getData();
+
+        // determine ordering
+        Map<String, Object> first = data.get(0);
+        Map<String, Object> last = data.get(data.size() - 1);
+        BigDecimal firstGroup = new BigDecimal(first.get(fieldNameKey).toString()); // assume number because 'number' fieldType
+        BigDecimal lastGroup = new BigDecimal(last.get(fieldNameKey).toString());
+        String order = "asc";
+        if (firstGroup.compareTo(lastGroup) == 1) {
+            order = "des";
+        }
+
+        // determine count
+        int count = 50; // default count
+        if (this.clusterClause.getCount() != 0) {
+            count = this.clusterClause.getCount();
+        }
+
+        // data is small enough
+        if (clusters == null && count > data.size()) {
+            return tabularQueryResult;
+        }
+
+        // find new clusters
+        List<Map<String, Object>> newData = getNewDataBins(fieldNameKey, firstGroup, lastGroup, count, clusters);
+
+        // take care of extra keys that may be present
+        Map<String, Object> extraKeySets = getExtraKeySetsMap(fieldNameKey, aggregationNameKey, data);
+
+        // aggregate data into the new clusters
+        aggregateNumbersInNewData(fieldNameKey, aggregationNameKey, data, newData, extraKeySets);
+
+        this.clusterClause = null;
```

Comment:
Does `this.clusterClause = null` need to be both here and in the `cluster` method?

Reply:
No, I've removed it in the next PR when I refactored the methods.

Reply:
Alright.